### PR TITLE
feat(payload): abstract payload builder in trait

### LINKED ISSUE #10956: Move `PayloadBuilderHandle` methods out to new trait
### Describe the feature

Move `PayloadBuilderHandle` methods out to new trait

https://github.com/paradigmxyz/reth/blob/0cde0724539962b80bc2eea390afbbfea501becd/crates/payload/builder/src/service.rs#L103-L200

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by greged93**
Abstracts the current `PayloadBuilderHandle` behind a `PayloadBuilder` trait. First step towards being able to have the `PayloadBuilder` be a GAT on the `NodeComponents` trait.

Resolves #10956.

### CODE REVIEW

File: crates/payload/builder/src/traits.rs [hunk 1]
```diff
@@ -1,9 +1,58 @@
...
 use reth_provider::CanonStateNotification;
-use std::future::Future;
+use std::{future::Future, pin::Pin};
+use tokio::sync::oneshot;
+
+pub(crate) type PayloadFuture<P> =
+    Pin<Box<dyn Future<Output = Result<P, PayloadBuilderError>> + Send + Sync>>;
+
+/// A type that can request, subscribe to and resolve payloads.
+#[async_trait::async_trait]
```
Comment (greged93) [inline crates/payload/builder/src/traits.rs:13]: I'm not too sure about this, are you ok with using the `async_trait` macro or would you rather returning a future?

Comment (mattsse) [reply crates/payload/builder/src/traits.rs:13]: async_trait is fine here, then we could use this as dyn trait as well.

File: crates/payload/builder/src/traits.rs [hunk 2]
```diff
@@ -1,9 +1,58 @@
...
+pub trait PayloadBuilder {
```
Comment (mattsse) [inline crates/payload/builder/src/traits.rs:14]: can we move this to payload-primitives perhaps?

Comment (mattsse) [inline crates/payload/builder/src/traits.rs:14]: this should be Send + Unpin

File: crates/payload/builder/src/service.rs [hunk 3]
```diff
@@ -141,64 +120,78 @@ where
...
+    async fn subscribe(&self) -> Result<PayloadEvents<Self::PayloadType>, Self::Error> {
         let (tx, rx) = oneshot::channel();
         let _ = self.to_service.send(PayloadServiceCommand::Subscribe(tx));
         Ok(PayloadEvents { receiver: rx.await? })
     }
 }
 
+impl<T> PayloadBuilderHandle<T>
+where
+    T: PayloadTypes + 'static,
```
Comment (mattsse) [inline crates/payload/builder/src/service.rs:159]: let's also make `PayloadTypes` 'static

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, some nits

**Comment by greged93** [comment]
I noticed there were 2 `PayloadBuilderError`, one in `reth-payload-builder` and one in `reth-payload-primitives`. They looked very similar barring the `MissingPayload` variant, so I removed the one from `reth-payload-builder`, and moved the events to primitives as well. This avoids any circular dep between builder and primitives.

File: crates/payload/builder/src/lib.rs [hunk 4]
```diff
@@ -28,7 +28,7 @@
 //! use std::pin::Pin;
 //! use std::task::{Context, Poll};
 //! use reth_payload_builder::{EthBuiltPayload, KeepPayloadJobAlive, EthPayloadBuilderAttributes, PayloadJob, PayloadJobGenerator};
-//! use reth_payload_builder::error::PayloadBuilderError;
+//! use reth_payload_primitives::PayloadBuilderError;
```
Comment (mattsse) [inline crates/payload/builder/src/lib.rs:31]: we should re-export the PayloadBuilderError in L125 then this wouldn't break too much and some crates don't require an additional dep

**Review by mattsse — CHANGES_REQUESTED** [review]
very nice, one pedantic nit about re-exports

**Review by mattsse — APPROVED** [review]

**Pull request merged**